src/ion/spill.rs

```diff
@@ -122,70 +125,40 @@ impl<'a, F: Function> Env<'a, F> {
                 self.slots_by_size.resize(
                     size + 1,
                     SpillSlotList {
-                        first_spillslot: SpillSlotIndex::invalid(),
-                        last_spillslot: SpillSlotIndex::invalid(),
+                        slots: smallvec![],
+                        probe_start: 0,
                     },
                 );
             }
             // Try a few existing spillslots.
-            let mut spillslot_iter = self.slots_by_size[size].first_spillslot;
-            let mut first_slot = SpillSlotIndex::invalid();
-            let mut prev = SpillSlotIndex::invalid();
+            let mut i = self.slots_by_size[size].probe_start;
             let mut success = false;
-            for _attempt in 0..10 {
-                if spillslot_iter.is_invalid() {
-                    break;
-                }
-                if spillslot_iter == first_slot {
-                    // We've started looking at slots we placed at the end; end search.
-                    break;
-                }
-                if first_slot.is_invalid() {
-                    first_slot = spillslot_iter;
-                }
+            for _attempt in 0..std::cmp::min(self.slots_by_size[size].slots.len(), MAX_ATTEMPTS) {
+                let spillslot = self.slots_by_size[size].slots[i];
```

Comment:
I found the correctness of this `[i]` oddly subtle even though it's ensured mostly by the line immediately above. It wasn't immediately obvious to me that the loop doesn't run when the slots array is empty. If you wanted to make my head hurt less, you could define a method on `SpillSlotList` that returns `Option<SpillSlotIndex>` and encapsulates the wrapping behavior that's currently split across several parts of this loop body. I don't think that's necessary for merging this fix, though.

Other than that: I'm convinced this patch does what you said it does, and it's nice that it deletes a bunch of code!

Reply:
Thanks, yes, it's not totally clear as-is! I went a bit further and defined an iterator; let me know what you think.

Reply:
I woke up in the middle of the night wondering whether probe_start is updated if the iteration limit is hit. I see that in either unsuccessful case you always reset it to the newly added slot. So that seems okay, I guess?

Thank you for the extended comments, that helps a lot!